google/cloud/storage/internal/openssl_util.h

```diff
@@ -0,0 +1,245 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_OPENSSL_UTIL_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_OPENSSL_UTIL_H_
+
+#include "google/cloud/internal/throw_delegate.h"
+#include "google/cloud/storage/internal/credential_constants.h"
+#include "google/cloud/storage/version.h"
+#include <openssl/bio.h>
+#include <openssl/buffer.h>
+#include <openssl/evp.h>
+#include <openssl/opensslv.h>
+#include <openssl/pem.h>
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+namespace google {
+namespace cloud {
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+/**
+ * Helper functions for Base64 and related transcoding.
+ */
+struct OpenSslUtils {
+ public:
+  /**
+   * Encode a string using Base64.
+   */
+  static std::string Base64Encode(std::string const& str) {
+    auto bio_chain = MakeBioChainForBase64Transcoding();
+    int retval = 0;
+
+    while (true) {
+      retval = BIO_write(static_cast<BIO*>(bio_chain.get()), str.c_str(),
+                         static_cast<int>(str.length()));
+      if (retval > 0) break;  // Positive value == successful write.
+      if (not BIO_should_retry(static_cast<BIO*>(bio_chain.get()))) {
+        std::ostringstream err_builder;
+        err_builder << "Permanent error in " << __func__ << ": "
+                    << "BIO_write returned non-retryable value of " << retval;
+        google::cloud::internal::RaiseRuntimeError(err_builder.str());
+      }
+    }
+    // Tell the b64 encoder that we're done writing data, thus prompting it to
+    // add trailing '=' characters for padding if needed.
+    while (true) {
```

Comment:
I tried to think of ways to avoid the `while (true)`, but I think the way you wrote these loops is the most readable way.

Reply:
Same.  I originally had a boolean in there, but it just added clutter and provided no extra semantic meaning.